optexp/datasets/wikitext103.py

```diff
@@ -0,0 +1,172 @@
+            def __init__(self):
+                self.sequences = sequences
+                self.targets = targets
+
+            def __len__(self):
+                return len(self.sequences)
+
+            def __getitem__(self, idx: int):
+                return self.sequences[idx], self.targets[idx]
+
+        return TorchDataset()
+
+    def get_tokens(self, tr_va: TrVa, vocab_size: int):
+        if self.is_tokenized(tr_va):
+            tokens = torch.load(
+                open(
+                    self.directory / f"wikitext103_{tr_va}_tokenized.pt",
+                    "rb",
+                )
+            )
+            return tokens
+        raw_data = "wiki.train.tokens" if tr_va == "tr" else "wiki.valid.tokens"
+        if not self.has_tokenizer():
+            self.tokenizer.build_tokenizer(
+                self.directory / f"wikitext103_v={vocab_size}",
+                self.directory / "wiki.train.tokens",
+                vocab_size,
+            )
+        tokens = self.tokenizer.tokenize_and_numify(
+            self.directory / f"wikitext103_v={vocab_size}.model",
+            self.directory / raw_data,
+        )
+        torch.save(
+            tokens,
+            open(
+                self.directory / f"wikitext103_{tr_va}_tokenized.pt",
+                "wb",
+            ),
+        )
+        return tokens
+
+    def is_tokenized(self, tr_va: TrVa):
+        return (self.directory / f"wikitext103_{tr_va}_tokenized.pt").exists()
+
+    def has_tokenizer(self):
+        return all(
+            (self.directory / file).exists()
+            for file in [
+                f"wikitext103_v={self.vocab_size}.model",
+                f"wikitext103_v={self.vocab_size}.vocab",
+            ]
+        )
+
+    def is_downloaded(self):
+        return all(
+            (self.directory / file).exists()
+            for file in [
+                "wiki.train.tokens",
+                "wiki.valid.tokens",
+                "wiki.test.tokens",
+            ]
+        )
+
+    def download(self):
+        os.makedirs(self.directory, exist_ok=True)
+        base_url = "https://huggingface.co/datasets/Salesforce/wikitext/resolve/main/wikitext-103-v1"
+        files = [
+            "validation-00000-of-00001.parquet",
+            "test-00000-of-00001.parquet",
+            "train-00000-of-00002.parquet",
+            "train-00001-of-00002.parquet",
+        ]
+        for file in files:
+            filepath = self.directory / Path(file)
+            with requests.get(f"{base_url}/{file}", stream=True) as r:
+                with open(filepath, "wb") as f:
+                    shutil.copyfileobj(r.raw, f)
+            data = pyarrow.parquet.read_table(filepath).to_pydict()
+            if file.startswith("train"):
+                split = "train"
+            if file.startswith("validation"):
+                split = "valid"
+            if file.startswith("test"):
+                split = "test"
```

Comment:
`split = file.split("-", maxsplit=1)[0]`?

Reply:
added sort of